Threaded validation

br648 opened the pull request:
### Checklist

- [ ] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [ ] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [ ] The description lists all applicable issues this PR seeks to resolve
- [ ] The description lists any configuration setting(s) that differ from the default settings
- [ ] All tests and CI builds passing

### Description

Working on top of https://github.com/ibi-group/datatools-server/pull/512/files original and mobility data validation are now in their own separate threads.

Commit:
refactor(threaded valiators): Original and mobility data validation executed in separate threads

br648 commented:
@binh-dam-ibigroup, Miles is to carry out some testing, but on the whole is happy to merge this into his working PR. If you don't mind, can you review please just in case I've missed something?

miles-grant-ibigroup commented on src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java:
```diff
@@ -382,22 +414,6 @@ public void validate(MonitorableJob.Status status) {
 
             // This will persist the document to Mongo
             this.mobilityDataResult = Document.parse(json);
```
The "regular" parsing step was previously persisting the `this` to mongo. If the mobility data parser finishes first, this still happens. If it finishes second, this fails to happen and the data isn't persisted! I think the solution is to add a mongo persistence line right here

br648 replied:
@miles-grant-ibigroup We need to be careful here because we don't want to create a new Mongo object if the other validator is expecting to do that. See ValidateFeedJob -> jobFinished(). Both validators need to be aware of the Mongo object state, so that either can create or update. Let me take a look. Good spot btw.

br648 replied:
@miles-grant-ibigroup This should now be addressed. Unit tests look good. Lemme know how you get on.

miles-grant-ibigroup reviewed (changes requested):
Seems to be working nice and fast, only I noticed to resolve is a race condition that sometimes prevents results from being persisted to mongo

Commit:
refactor(Handling of feed version saving): Control saving and update of feed version in relation to

dependabot[bot] committed:
chore(deps): bump mongodb-driver-sync from 4.0.5 to 4.0.6

Bumps [mongodb-driver-sync](https://github.com/mongodb/mongo-java-driver) from 4.0.5 to 4.0.6.
- [Release notes](https://github.com/mongodb/mongo-java-driver/releases)
- [Commits](https://github.com/mongodb/mongo-java-driver/compare/r4.0.5...r4.0.6)

---
updated-dependencies:
- dependency-name: org.mongodb:mongodb-driver-sync
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

binh-dam-ibigroup committed:
Merge pull request #475 from ibi-group/dependabot/maven/org.mongodb-mongodb-driver-sync-4.0.6

chore(deps): bump mongodb-driver-sync from 4.0.5 to 4.0.6

Commit:
Merge branch 'mobility-data-validator' into threaded-validation

Commit:
refactor(Added feed version delete): Delete the feed version if the ValidateMobilityDataFeedJob fail

binh-dam-ibigroup commented on src/main/java/com/conveyal/datatools/manager/jobs/ValidateFeedJob.java:
```diff
@@ -44,13 +44,7 @@ public void jobFinished () {
-                } else {
-                    Persistence.feedVersions.replace(feedVersion.id, feedVersion);
-                }
+                feedVersion.persistFeedVersionAfterValidation(isNewVersion);
```
As written, it looks like the new version is shown in the UI (on the `mobility-data-validator` branch) after the first validator completes. Is that the desired effect?

miles-grant-ibigroup replied:
I think that's ok as long as we refresh the UI when the mobility data results come in, which is something we can do on the client.

miles-grant-ibigroup committed:
Merge branch 'dev' into threaded-validation

miles-grant-ibigroup committed:
chore(Dockerfile): update java version

miles-grant-ibigroup reviewed (approved):
Things are working well for me, I think this is good to merge back into the main mobility data branch thanks for the big improvement!

miles-grant-ibigroup commented:
I've noticed that on some large feeds the jobs are running sequentially?
<img width="242" alt="Screenshot 2023-03-01 at 11 50 35 AM" src="https://user-images.githubusercontent.com/86619099/222207288-29ceaebf-d9da-4cd2-ac7b-b0cdf884428e.png">

br648 commented:
This is to be expected. Once both threads have started we are then effectively passing control to the JVM which then decides how much CPU each thread will get. If the first gets 100% CPU, the second will have to wait.

binh-dam-ibigroup reviewed (approved):
Good deal!
